### PR TITLE
[FlexAttention] Skip Calculating KV Gradient if Possible

### PR DESCRIPTION
**Pull request by cora-codes**
Partial implementation of #136101

This tackles the easy case. Adding additional support for skipping the query gradient would be trivial, modifying flex attention to just calculate the key or value's gradient is possible but more involved.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by cora-codes** [comment]
In the near future, it would nice to extend `BlockMask` to lazily (?) require/calculate `q_num_blocks`, `q_indices`, `full_q_num_blocks` and `full_q_indices` since they are only required to calculate `dKV`!

**Review by drisspg — COMMENTED** [review]
Can you write a Unit test?
As well could you gather some perf numbers w/ and w/out kv requring grad
can you write some tests, as well would you measuring to see if there is a meaningful perf impact

thanks for putting up the PR

File: torch/_inductor/kernel/flex_attention.py [hunk 1]
```diff
@@ -1803,6 +1807,11 @@ def flex_attention_backward(*args, **kwargs):
         kernel_options.setdefault("SPARSE_Q_BLOCK_SIZE", SPARSE_Q_BLOCK_SIZE)
         kernel_options.setdefault("SPARSE_KV_BLOCK_SIZE", SPARSE_KV_BLOCK_SIZE)
 
+        if k_requires_grad and v_requires_grad:
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:1810]: Or?

**Comment by cora-codes** [comment]
Right now, the performance seems quite poor. I imagine it is because it is parallelized along `batch * kv_heads`. If you have any ideas for how to improve it - I'm happy to try them, but I'm not sure this is worth pushing forward as is...

**Comment by tvaranka** [comment]
> In the near future, it would be nice to extend `BlockMask` to lazily (?) require/calculate `q_num_blocks`, `q_indices`, `full_q_num_blocks` and `full_q_indices` since they are only required to calculate `dKV`!
> 
> One more thought - the most elegant solution may not be adding more logic into the flex attention kernel, but instead allowing for additional control and flexibility in the backward pass. For instance, we could allow users to pass a function that modifies the forward `BlockMask` to produce different sparsity in the backward pass. This would not only support the default `_dense_to_ordered`-like behavior but also enable:
> 
> * Skipping the calculation of Q, K or V gradients
> * Data-dependent sparsity in the backward pass
> * Ignoring specific sections of context in the backward pass (useful for masking out copyrighted documents during training, for instance)
> 
> This approach could provide a more versatile and extensible solution without significantly increasing the complexity of the core kernel implementation.
> 
> (In theory, you can do something similar today by constructing your own `q_num_blocks` and `q_indices`... however, I experimented with that strategy for not calculating KV gradient and it didn't seem to work how I expected)

Interested in masking the backward pass. Is there any way to do this currently?

**Comment by Chillee** [comment]
@tvaranka wdym by "masking" the backwards pass? You want a different mask in the backwards than the forwards?

**Comment by tvaranka** [comment]
> @tvaranka wdym by "masking" the backwards pass? You want a different mask in the backwards than the forwards?

Yeah, I am not sure if that makes sense or if this would be a solution for it, but basically avoid updating some combinations of Q and K when fine-tuning. So, after fine-tuning the attention map for certain locations wouldn't be changed.
